retry/retry.go

```diff
@@ -91,19 +91,27 @@ func (r Retry) waitTimeInNs(attempt int) int {
 
 // WaitTimeInNs calculates the wait time in nanoseconds based off the attempt number based off
 // exponential backoff with a random delay. It caps at the constant maxWaitTimeInNs.
-func WaitTimeInNs(attempt int, random *rand.Rand) int {
+func WaitTimeInNs(attempt int, random *rand.Rand) time.Duration {
 	a := float64(attempt)
+
+	// Check if max attempts reached, by reversing the calculation
+	// for the wait time using the maxWaitTime
+	maxBaseTimeWithDelayInS := float64(maxWaitTimeInNs) / float64(time.Second)
+	maxA := (math.Log(2 * maxBaseTimeWithDelayInS / 3)) / math.Log(2)
+	if a >= maxA {
+		return maxWaitTimeInNs
+	}
```

Comment:
Curious about this new addition to calculate the number of attempts needed for the maxBaseTimeWithDelayInNS. Was there an issue with the previous calculation method? I found the previous way of handling this easier to read

Reply:
With the change to duration, it became apparent that the attempt number passed in overflows really fast (like 30 attempts) due to the exponents and conversion to nanoseconds. So I added this calculation to eliminate this. I'll actually also add a check for the max int to make sure we don't ever overflow

Reply:
Introduced this since I noticed the attempts was overflowing at really low integers (like 30) due to the exponents and converting to nanoseconds. This makes it so we can accept the full range of integers.